ocaml/typing/typedtree.mli

```diff
@@ -406,11 +406,14 @@ and 'k case =
    [Guard_pattern] *)
 and guard =
   | Predicate of expression
-  | Pattern of expression * Layouts.sort * computation general_pattern
-  (* [Pattern (scrutinee, sort, pattern)] represents a pattern guard. The case
-     will be taken if [scrutinee] evaluates to a value matching [pattern].
-     Variables bound by the pattern match are available on the RHS of the case.
-     Like the [Pexp_match] constructor, [sort] is the sort of the scrutinee. *)
+  | Pattern of expression * Layouts.sort * computation general_pattern * partial
```

Comment:
When we hit six fields, and probably even before that, it's a good idea to consider whether a record is a better fit than a tuple. In the compiler codebase, records have the downside that (because warning 9 is disabled) you aren't encouraged to write patterns that enumerate all the fields, so this truly is a tradeoff, but here I think it's fine.

(This is a little deep in the details, but you can locally enable warning 9 for the duration of a let/match if you're interested in having exhaustivity again. That's: `match[@ocaml.warning "+9"] blah with ...`. `ast_iterator` is the main use case for something like this.)

A recommendation: be thoughtful with the record fields' names. A good name could make clear, for example, that the sort belongs to the expression and not the pattern, and could identify what the env corresponds to (is it the env from before or after binding the pattern's variables?) I'm happy to brainstorm this with you before you go through and update all the use sites of these names.

Reply:
I did update this, but if the names could be more precise, we can update them again (it's not too much work).